client/client_test.go

```diff
@@ -2196,6 +2197,49 @@ func testBuildHTTPSource(t *testing.T, sb integration.Sandbox) {
 	require.Equal(t, http.MethodHead, allReqs[1].Method)
 	require.Equal(t, "gzip", allReqs[1].Header.Get("Accept-Encoding"))
 
+	require.NoError(t, os.RemoveAll(filepath.Join(tmpdir, "foo")))
+
+	// update the content at the url to be gzipped now, the final output
```

Comment:
The problem happens when [there is a cache metadata match on the URL hash](https://github.com/moby/buildkit/blob/333ee9158128dbb821118f0b19a4c140d49c7657/source/http/httpsource.go#L179) but then it turns out the actual URL (as checked with `HEAD`) does not have the same etag as what was in cache. That triggers the codepath where `HEAD` gets the explicitly set `Accept-Encoding` (fine by itself) which is [then retained for the subsequent `GET`](https://github.com/moby/buildkit/blob/333ee9158128dbb821118f0b19a4c140d49c7657/source/http/httpsource.go#L207) and causes the file to not be transparently decompressed by the go stdlib (the actual bug fixed here)

Wrote the test like this because it seemed like a realistic scenario (URL changes contents), but the reason we hit it in Dagger is slightly different. We were always setting `llb.Filename` to the same value for every `llb.HTTP` source, (due to the fact that it always gets mounted/copied to a different name once utilized by users, so the name was internal-only and didn't really matter).

We didn't realize this meant that unless the user set non-default uid/gid/perms, the [url hash](https://github.com/moby/buildkit/blob/333ee9158128dbb821118f0b19a4c140d49c7657/source/http/httpsource.go#L91) would always be the same because [the filename is used there instead of the url](https://github.com/moby/buildkit/blob/333ee9158128dbb821118f0b19a4c140d49c7657/source/http/httpsource.go#L415-L417).

So, essentially every single one of our HTTP sources ended up with the same url hash, which meant they all got the same match [here](https://github.com/moby/buildkit/blob/333ee9158128dbb821118f0b19a4c140d49c7657/source/http/httpsource.go#L134). This meant that after the first HTTP source was used, we'd almost always trigger the codepath where we call `HEAD`, it doesn't match the cached etag, we call `GET`, that request doesn't get decompressed if it's gzipped.

I'm also fixing Dagger to use a hash of the URL as the filename instead so we can have more optimal behavior here, but that's independent of the bug w/ not decompressing the response, which can also happen in fairly straightforward scenarios like the test case here where the URL just changes contents.

Reply:
Just looked into this, the problem is that if you set `Checksum` then most of the codepaths relevant here are skipped: https://github.com/moby/buildkit/blob/333ee9158128dbb821118f0b19a4c140d49c7657/source/http/httpsource.go#L123-L126

The existing test asserts on the content by exporting the result and looking at the contents of the file there.